source/sio/epoll/epoll_context.hpp

```diff
@@ -0,0 +1,897 @@
+/*
+ * Copyright (c) 2023 Xiaoming Zhang
+ *
+ * Licensed under the Apache License Version 2.0 with LLVM Exceptions
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *   https://llvm.org/LICENSE.txt
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <bits/chrono.h>
+#include <cerrno>
+#include <concepts>
+#include <cstdint>
+#include <sys/epoll.h>
+#include <sys/eventfd.h>
+#include <sys/timerfd.h>
+#include <sys/uio.h>
+
+#include <atomic>
+#include <chrono>
+#include <cstring>
+#include <optional>
+#include <system_error>
+#include <type_traits>
+#include <utility>
+
+#include "exec/__detail/__atomic_intrusive_queue.hpp"
+#include "exec/linux/safe_file_descriptor.hpp"
+#include "exec/scope.hpp"
+#include "stdexec/execution.hpp"
+#include "stdexec/__detail/__execution_fwd.hpp"
+#include "stdexec/stop_token.hpp"
+#include "sio/assert.hpp"
+#include "exec/__detail/__manual_lifetime.hpp"
+#include "exec/timed_scheduler.hpp"
+
+namespace sio {
+  namespace epoll {
+    inline void throw_error_code_if(bool cond, int ec) {
+      if (cond) {
+        throw std::system_error(ec, std::system_category());
+      }
+    }
+
+    inline int create_epoll() {
+      static constexpr int epoll_create_size_hint = 1024;
+      int fd = ::epoll_create(epoll_create_size_hint);
+      throw_error_code_if(fd < 0, errno);
+      return fd;
+    }
+
+    inline int create_eventfd() {
+      int fd = ::eventfd(0, EFD_CLOEXEC | EFD_NONBLOCK);
+      throw_error_code_if(fd < 0, errno);
+      return fd;
+    }
+
+    inline int create_timer() {
+      int fd = ::timerfd_create(CLOCK_MONOTONIC, 0);
+      throw_error_code_if(fd < 0, errno);
+      return fd;
+    }
+
+    struct context_base : stdexec::__immovable {
+      context_base()
+        : epoll_fd_(create_epoll())
+        , event_fd_(create_eventfd()) {
+        epoll_event event = {.events = EPOLLIN | EPOLLERR | EPOLLET, .data = {.ptr = &event_fd_}};
+        int ret = ::epoll_ctl(epoll_fd_, EPOLL_CTL_ADD, event_fd_, &event);
+        throw_error_code_if(ret < 0, errno);
+      }
+
+      ~context_base() {
+        epoll_event event = {};
+        int ret = ::epoll_ctl(epoll_fd_, EPOLL_CTL_DEL, event_fd_, &event);
+        throw_error_code_if(ret < 0, errno);
+      }
+
+      exec::safe_file_descriptor epoll_fd_;
+      exec::safe_file_descriptor event_fd_;
+    };
+
+    struct operation_base;
+
+    struct operation_vtable {
+      // return true means operation is done and associated context will complete operation.
+      bool (*execute_)(operation_base*) noexcept = nullptr;
```

Comment:
I really wonder if that's enough but its a good start. I was surprised to see only two methods :-)